### ISSUE TITLE
Long node labels mess up self-regulatory arrows

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#318]
Long node labels mess up self-regulatory arrows in terms of their placement and the arrowhead alignment on the edge.
![long-node-label-autoregulatory-arrow](https://cloud.githubusercontent.com/assets/6463867/17447143/b4539d70-5b01-11e6-99bd-5ffe43f7ef70.png)

Probably need to also reduce whitespace on edges of label.

Low priority.

**Comment by kdahlquist** [comment]
Assigning this one to @yshin4 to work on after she finishes with #333.

**Comment by yshin4** [comment]
Noticed even shorter gene name nodes (length 4) has an arrow that is off as well when it is bold, will look more into it once the #333 is finished.

**Comment by yshin4** [comment]
@kdahlquist 

Does this mean you would like the fixed graph to look like the top node, bottom node, or in some other form?

![image](https://cloud.githubusercontent.com/assets/16246811/24938489/59c53652-1eec-11e7-8069-94cb1dc02c0a.png)

**Comment by kdahlquist** [comment]
@yshin4, not really either of those drawings.  The self-regulatory arrows should look like they do now for smaller nodes, like this:
![self-regulatory-edge-long](https://cloud.githubusercontent.com/assets/6463867/24978434/89a13ff8-1f85-11e7-8e9b-0a7db0410532.jpg)

**Comment by kdahlquist** [comment]
@yshin4, this is a higher priority than either #392 or #188.  Both of those are in a better state than this one.  See if you can make some headway on this one for the release and we can come back to the other ones later, if need be.

**Comment by yshin4** [comment]
Got an error and can't upload any files to the website on my fork server. #455

**Comment by kdahlquist** [comment]
I'm de-labeling this "for next release" because, presumably, @yshin4 won't be able to do any further work on this for the semester.

**Comment by yshin4** [comment]
Yes, sorry! I got hit by five final exams and one project. My apologies for not being able to make any progress.

**Comment by yshin4** [comment]
Taking a look at graph.js where getSelfReferringEdge is being set. My original plan was to change the x value of the self referring edge when the node is long (larger than 150px box/text width) but can't find a way to move the x position. Currently investigating why SELF_REFERRING_Y_OFFSET works but changing values for selfReferringEdgeWidth doesn't work.

**Comment by yshin4** [comment]
Had a mini session with dondi and was able to figure out which part of the code to look at in order to fix the edge locations (x,y) with his help. Added shifting value for x1, which is the x-position of where the edge starts and added shifting value for x2, which is the x-position of where the edge ends, when the node width is too large.

![image](https://user-images.githubusercontent.com/16246811/30181927-d7b5c502-93c9-11e7-960a-fb68e4e70159.png)

This is how it looks after the change is applied. @kdahlquist please see if there are any necessary changes to be made in addition.

**Comment by kdahlquist** [comment]
This looks good.  The point of the arrowhead is just a tiny bit underneath the box--I don't know if that can be fixed.  Can you share a screenshot of how it looks like with weighted edges?  You probably need to make a new test file (or several) because there are a lot of edge thickness X node size possibilities.

**Comment by yshin4** [comment]
Will be getting back to the issue Tuesday noon. I feel like the tip of the arrowhead being slightly covered by the node is related to issue #392, but not confirmed. I will also post a screenshot of how it looks with weighted edges before I start working on it tomorrow.

**Comment by yshin4** [comment]
![weighted](https://user-images.githubusercontent.com/16246811/30307235-888d0e7a-9731-11e7-814c-f5ffb6b53357.PNG)

This is how it looks with some of the weighted edge possibilities. I was mostly checking for the longer nodes (wwwwwwwwwwww and more ws) with thicker edges and it seems like the 'Long node labels mess up self-regulatory arrows' is not happening. I can make more tests if @kdahlquist wants, but if this looks good enough I will move on to working on the arrowheads being covered.

**Comment by yshin4** [comment]
Posted a merge request.

**Comment by kdahlquist** [comment]
This is looking good.  I've looked at beta 2.2.2.  However, has the test file with the i's and w's that you show a screenshot of in a previous post been committed to the repository?  If so, I can't find it.  Whatever test files were used to test this should be committed.  I won't close this until this has been confirmed.

**Comment by kdahlquist** [comment]
I'm looking on the beta branch and I can't easily see where the test files are.  Which directory are they in?  Shouldn't they be in edge tests?

I do want to close this, but I want to verify that the test files are preserved.

**Comment by yshin4** [comment]
@kdahlquist It was because my pull request was not approved until about an hour ago. They should be in edge-tests if you check now.

**Comment by kdahlquist** [comment]
In the file all-negative-weights, the blunt end markers between E-11 to E-12 and E-13 to E-14 are not centered when the edges are in the vertical direction up or down.

In the file check-self-referential-widths the blunt market for S-14 Is not centered.

@dondi and I discussed that if you fix this during your meeting tomorrow and he reviews it then I don't need to do additional review.

**Comment by yshin4** [comment]
Will do.

**Comment by yshin4** [comment]
Mentioned markers are centered and @dondi cofirmed. However, we found an issue with left vertical distance being too large. We discussed a formula to fix the issue and I will be implementing this on Monday during my session with eddie & eileen.

**Comment by yshin4** [comment]
The formula goes: 
u = (r-x1) / (x2-x1) 
r = targetNode.x + node.length/2 
x1,y1 = targetNode
x2,y2 = sourceNode
y = (1-u)y1 + uy2
Condition is if (u>0 && y>= y1 + 15 && y <= y1 - 15){
we apply left distance
}

**Comment by yshin4** [comment]
Formula applied, and it looks great @dondi!

![image](https://user-images.githubusercontent.com/16246811/31647478-4eafae5e-b2bd-11e7-8b1c-a1c19cd0903c.png)

![image](https://user-images.githubusercontent.com/16246811/31647494-725145ca-b2bd-11e7-97e4-83e3f5b71fef.png)

![image](https://user-images.githubusercontent.com/16246811/31647527-b7163468-b2bd-11e7-99dd-27e26fa945ec.png)

![image](https://user-images.githubusercontent.com/16246811/31647550-d6320854-b2bd-11e7-8a2e-c5365f78563e.png)

**Comment by kdahlquist** [comment]
@yshin4 are you talking to me or @dondi?  This looks fine.  E-7 to E-8 has only been shown in the vertical orientation in your screenshots, but can I assume you looked at them in the left and right orientation?

Has this code been pushed?

**Comment by yshin4** [comment]
@kdahlquist yes I did, I put them in a straight line but E-14 didn't fit so I decided to align it like how it looks above. I was talking to @dondi about the formula that we discussed to check if I implemented it right. However, since it looks good to @kdahlquist I will make a pull request. The code has been pushed to my branch but not beta yet.

**Comment by kdahlquist** [comment]
OK.  I wasn't actually talking about E-14.  E-7 to E-8 is being shown in the vertical orientation in all four screenshots.  I was just double-checking that you looked at that particular edge in the left and right orientations.

**Comment by kdahlquist** [comment]
OK.

**Comment by dondi** [comment]
Hi @yshin4 and @kdahlquist —Sorry for hopping on late; I missed the initial notification last night and have been in class this morning.

Those edges definitely look good, and better than they have so far! Glad that the equation I showed you last week worked out nicely.

I saw your pull request; Travis is unhappy with it but I'm not sure what to do about it. One unit test is failing because it is timing out and the other issue is that this makes our coverage go down. I’ll have to think about how to resolve those.

But meanwhile, I just have a couple of sanity-check notes:
* Did you try this code out with varying name lengths? I just want to make sure that things still look good when the node widths vary.
* I don't consider this a blocker but if it's easy to fix then I'd like to explore it. I know that the distance of the blunt marker from the node is partially a function of its thickness, and that's why it's been hard to get this right. But given that we now have an approach to precisely know when the marker is on the right side (and by extension, if needed we can do similar calculations for markers on the top, left, and bottom), can we tweak the distances so that all thicknesses now have the same visual distance from the node? For example, in the very top graph in https://github.com/dondi/GRNsight/issues/318#issuecomment-337118547, there appears to be a jump in distance (to my eye at least) when going from E-8 to E-9 vs. going from E-9 to E-10 (the distance of the blunt marker from E-9 to E-10 seems disproportionately farther than the marker is from E-8 to E-9).
  Is there a way to adjust this now so that the distances either stay uniform (i.e., we compensate for the thickness) or at the very least, have the distance scale up in a smoother way? Again, I don't think this is a blocking issue, but if we do a quick assessment on how easy or hard this might be, I would consider this to be a final loose end for this whole set of issues.

Otherwise, looks good, this is definitely a nice milestone!

**Comment by yshin4** [comment]
Yes, I noticed that the coverage slightly went down as well. That's a good point, I will make a long-node-negative test file and check. You are right about the distance changing from E-9. I will adjust the values so that the distance looks as similar as possible, if not the same. Unfortunately, I have a quiz tomorrow that I have to study for so I will not be able to make progress until tomorrow meeting. Thank you for the suggestions @dondi.

**Comment by dondi** [comment]
Three loose ends before we can move on:
* Fix the `node length` constant variable.
* Check blunt marker behavior when node names are long.
* Even out distance from E-8/E-9/E-10.

**Comment by yshin4** [comment]
![image](https://user-images.githubusercontent.com/16246811/31744318-3aaadf3a-b413-11e7-8285-de943b6c21fa.png)

I made some changes and now they look pretty even to me, what do you say @dondi?

**Comment by yshin4** [comment]
![image](https://user-images.githubusercontent.com/16246811/31744893-94d98964-b415-11e7-9409-e2c1964e61fc.png)

Made a test for long nodes, we can maybe take a more detailed look during office hours tomorrow.

**Comment by dondi** [comment]
Completed, with test files, and now on the `beta` branch.
